### PR TITLE
ENH: maybe_convert_objects corner cases

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -2550,6 +2576,26 @@ def maybe_convert_objects(ndarray[object] objects,
     return objects
 
 
+cdef _infer_all_nats(dtype, ndarray datetimes, ndarray timedeltas):
+    """
```
Comment (jreback) [inline pandas/_libs/lib.pyx:2580]: can you type dtype & narrow the types on datetimes/timedeltas if possible.

also a return type ?

File: pandas/tests/dtypes/test_inference.py [hunk 2]
```diff
@@ -665,6 +665,45 @@ def test_maybe_convert_objects_datetime(self):
...
+        if dtype == "timedelta64[ns]":
+            stamp = stamp - datetime(1970, 1, 1)
+        arr = np.array([stamp], dtype=object)
+
+        out = lib.maybe_convert_objects(
+            arr, convert_datetime=True, convert_timedelta=True
+        )
+        # no OutOfBoundsDatetime/OutOfBoundsTimedeltas
+        tm.assert_numpy_array_equal(out, arr)
+
```
Comment (jreback) [inline pandas/tests/dtypes/test_inference.py:706]: can you exercise the ValueError in infer_all_nats

**Review by jreback — CHANGES_REQUESTED** [review]
does this have any user facing change?

**Comment by jbrockmendel** [comment]
> does this have any user facing change?

not that im aware of.

motivating goal is to eventually kill off maybe_infer_to_datetimelike (along with the PR deprecating the string inference) and just use `lib.maybe_convert_objects(values, convert_datetime=True, convert_timedelta=True)`

File: pandas/_libs/lib.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/lib.pyx:2580]: not really, as there is no good way to declare `ndarray[dt64]`, and both the dtype arg and return type aren't representable here

File: pandas/tests/dtypes/test_inference.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/dtypes/test_inference.py:718]: sure, just pushed

**Comment by jreback** [comment]
can you rebase.

maybe_convert_objects is getting pretty long......

**Comment by jbrockmendel** [comment]
> maybe_convert_objects is getting pretty long......

on the plus side its mostly flat (at least the recent additions are).  im optimistic we can get rid of some other conversion functions/special-casing and it'll be extremely worth it (xref #40489 could use your input)

**Review by jreback — APPROVED** [review]

**Pull request merged**